### ISSUE TITLE
GPU Underutilized

### ISSUE THREAD

**Issue by iprovalo** [tensorflow/tensorflow#46800]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): NO
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04.5 LTS
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): v2.4.0-rc4-71-g582c8d236cb 2.4.0
- Python version: 3.7.9
- CUDA/cuDNN version: CUDA Version: 11.0 
- GPU model and memory: Tesla T4, 15gb

**Describe the current behavior**
Sample code on a 1-GPU is utilizing 20% GPU.  Increasing the number of GPU's lowers utilization per GPU and increases the training total time.  There are a lot of copying from H2D and back.  Some functions, like Identity (which maybe expected) are executed on CPU. 

1-GPU:
Epoch 2/2
1000/1000 [==============================] - 4s 4ms/step

8-GPU:
Epoch 2/2
1000/1000 [==============================] - 8s 8ms/step

**Describe the expected behavior**
GPU utilization should be much higher.  I am already using TF_GPU_THREAD_MODE=gpu_private option.  
The dataset is preprocessed and saved on to a snapshot, then loaded as a distributed dataset.  
The actual prod data is displaying similar characteristics to this sandbox test.

**Standalone code to reproduce the issue**
https://colab.research.google.com/drive/1C5gK_TshqO-fscJYIRDI6OX74b2l2Iw7

**Other info / logs** Include any logs or source code that would be helpful to
https://github.com/iprovalo/tf/blob/main/input_pipeline_stats_benchmark.py

**Comment by iprovalo** [comment]
Updated the code with a more realistic scenario with more features, batch size 100, 100K samples.  GPU up to 20%, but similar patterns can be seen in the profiler.

![Screen Shot 2021-01-30 at 9 04 38 AM](https://user-images.githubusercontent.com/1660997/106362998-b2f93080-62da-11eb-8bd0-60fa2f2374e3.png)
![Screen Shot 2021-01-30 at 9 04 26 AM](https://user-images.githubusercontent.com/1660997/106363001-b55b8a80-62da-11eb-9dd9-78e15d38c119.png)
![Screen Shot 2021-01-30 at 9 04 19 AM](https://user-images.githubusercontent.com/1660997/106363002-b55b8a80-62da-11eb-9859-15af1cace9dd.png)
![Screen Shot 2021-01-30 at 9 03 57 AM](https://user-images.githubusercontent.com/1660997/106363003-b5f42100-62da-11eb-9dbe-2c65d49ce0c0.png)

**Comment by iprovalo** [comment]
8-GPU:
![Screen Shot 2021-01-30 at 9 21 56 AM](https://user-images.githubusercontent.com/1660997/106363343-d624df80-62dc-11eb-9e58-85c57545cd6a.png)
![Screen Shot 2021-01-30 at 9 21 50 AM](https://user-images.githubusercontent.com/1660997/106363344-d6bd7600-62dc-11eb-9f20-48be9e5347f6.png)

**Comment by iprovalo** [comment]
Increasing the batch size to 1K improves the GPU utilization slightly, maybe 5%.  The overall pattern seems to remain the same.

**Comment by iprovalo** [comment]
@psobot I think this is possibly related to the issue you describe in [44194](https://github.com/tensorflow/tensorflow/issues/44194).

I have tried your solution, but didn't notice any improvements for our app.  However, I am observing some ops on the CPU, which I am not sure are supposed to be there.  The TF ops on the timeline look a bit too spread out and there seems to be a lot of H2D copying.  

I have a small [reference](https://github.com/iprovalo/tf/blob/main/input_pipeline_stats_benchmark.py) app to profile these issues.  I didn't include the Embeddings in my reference app.

**Comment by nikitamaia** [comment]
Hi @iprovalo, have you had a chance to look at the [Optimize GPU Performance](https://www.tensorflow.org/guide/gpu_performance_analysis) guide?

**Comment by iprovalo** [comment]
@nikitamaia yes I have used the suggestions from this guide.

This is a bare bones benchmark, no metrics, preprocessed input is read from the snapshot.  

My main question is why there are still so many ops running on cpu?  I think it could be creating the kernel launching bottlenecks.

Does tf have something similar to this benchmark to demo an end-to-end optimized workflow described in the guide?

**Comment by iprovalo** [comment]
Updated the benchmark code with embedding layer with generated data.

**Comment by iprovalo** [comment]
I added two more configs - Dense layer dimension and a number of layers to make the model more complex.  https://github.com/iprovalo/tf/blob/main/input_pipeline_stats_benchmark.py#L32-L33

With Dense layer dim=1024 and using five layers instead of one, I can get close to 80% utilization (with 10 layers, for sure 100%).  

With more complex set up, the model's kernel launch time decreases relatively to the processing time, it's around 10% which sounds much better and a GPU profile in the tracer (see attached) looks much different, but at the same time, the absolute processing time per step is growing from 7.6ms to 48.2ms when I run it on:

1-GPU (40% utilization, 7.6ms)
2-GPU (50% utilization, 11.4ms)
4-GPU (57% utilization, 23.3ms)
8-GPU (80% utilization, 48.2ms)

If I try The MultiWorkerMirroredStrategy suggested as a work-around in https://github.com/tensorflow/tensorflow/issues/41898, I get 41.7ms per step (8-GPU).  It's not helpful since it still increases the time per step by 6X.  With tf.distribute.MirroredStrategy(cross_device_ops=tf.distribute.HierarchicalCopyAllReduce()) or ReductionToOneDevice the timing goes to ~52ms (8GPU). 

So, the overall timing is getting worse with the addition of the GPU's (~7X for 1GPU->8GPU).  If I select top ~10 Kernels in the profiler, this reduce op takes 95%: ncclAllReduceRingLLKernel_sum_f32.  The long line in the attached tracer is this reducer.

Related issues:
https://github.com/tensorflow/tensorflow/issues/39545
https://github.com/tensorflow/tensorflow/issues/41898

The work-around suggested in the https://github.com/tensorflow/tensorflow/issues/41898 is not working for this case.

![Screen Shot 2021-02-18 at 7 52 22 AM](https://user-images.githubusercontent.com/1660997/108391558-ae7fb380-71c6-11eb-9e0a-3779d1d6d20a.png)

**Comment by okurman** [comment]
I'm facing the same issue. Can't get distributed training running. It is about 2x slower than a single GPU training...

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
